src/main/java/dev/morling/onebrc/CalculateAverage_cliffclick.java

```diff
@@ -0,0 +1,394 @@
+    static void do_chunk(Work w, boolean skip1, int limit, MappedByteBuffer mmap) {
+        assert mmap.isDirect();
+        int idx = 0;
+        int max = mmap.limit();
+        // If start>0, skip until first newline
+        if (skip1) {
+            while (mmap.get(idx) != '\n')
+                idx++;
+            // WINDOWS
+            // idx++;
+        }
+
+        // For this chunk of file do...
+        while (idx < limit) {
+            int cityx = idx; // Used if we find a new city name
+
+            // SWAR read and build n8; the long-as-a-string value. Also track start
+            // and end of the string, in case it is new and needs to be inserted into
+            // the n8->city_name map.
+            long n8 = 0;
+            // Read a misaligned long
+            long x = mmap.getLong(idx);
+            // Found semi ?
+            long hasM = has0(x ^ HASSEMI);
+            while (hasM == 0) {
+                // Read 2nd word of city
+                n8 ^= x;
+                idx += 8;
+                // Edge case for last few bytes in mmap
+                if (idx + 8 > max) {
+                    x = mmap.getLong(max - 8);
+                    int skip = (8 - (max - idx));
+                    idx -= skip; // Roll back skip amount
+                    x = x & (-1L >>> (skip << 3)); // Get just the last few bytes in mmap
+                }
+                else {
+                    // Read a misaligned long
+                    x = mmap.getLong(idx);
+                }
+                // Found semi ?
+                hasM = has0(x ^ HASSEMI);
+            }
+            // Found a semicolon this word.
+            // There is only one, because city names are not that short.
+            // The high bit of the byte in question is set.
+            int semidx = Long.numberOfTrailingZeros(hasM);
+            int shr = semidx + 1;
+            if (shr != 64) {
+                n8 ^= (x >> shr);
+                idx += 8 - ((semidx + 1) >> 3);
+            }
+
+            // Skip semicolon
+            idx++;
+
+            // Reading tempature:
+            int temp = 0;
+            boolean neg = false;
+            byte b = mmap.get(idx++);
+            if (b == '-') {
+                neg = true;
+                b = mmap.get(idx++);
+            }
+            temp = b - '0';
+            b = mmap.get(idx++);
+            if (b != '.') {
+                temp = temp * 10 + b - '0';
+                idx++;
+            }
+            // Read fraction digit; scaled decimal temp
+            b = mmap.get(idx++);
+            temp = temp * 10 + b - '0';
+            if (neg)
+                temp = -temp;
+            // Skip newline
+            idx++;
+            // F*KING WINDOWS.
+            // Skip CR
+            // idx++;
+
+            w.insert(n8, temp, mmap, cityx);
+        }
+    }
+
+    private static class Work {
+        private static final int TAB_SIZE = 0x800; // 512 for 413 cities
```

Comment:
Is this a "legal" assumption according to the rules? 🤔

Reply:
Yes and no; it fails the 10000 city test case but increasing the table size to 16K fixes that.  Other than that, there's nothing in the rules about supporting more and more cities until memory is exhausted - so a resizing hash table is strictly not required.  If the test cases forced e.g. a million entry table, but the performance work only required a 512 entry table then it might make sense to add the resize logic - as have such a large diffuse table would probably impact caches, and hurt the performance test.

Reply:
Yes, this looks ok now. To be valid, it must pass the 10K city test (which it does now), and it must pass evaluation (i.e. yield the expected output) for the original 400 cities key set.